### PR TITLE
feat(api)!: pass args table to autocommand callbacks

### PR DESCRIPTION
**Pull request by gpanders**

### CODE REVIEW

**Comment by fsouza** [comment]
This is a breaking change in Lua too. Example:

```lua
local function do_thing(bufnr)
  bufnr = bufnr or vim.api.nvim_get_current_buf()
  -- do things
end

vim.api.nvim_create_autocmd(["BufNewFile"], { callback = do_thing })
```

**Comment by lewis6991** [comment]
The arguments in a callback shouldn't ever have been relied on so it's more soft breaking.

**Comment by clason** [comment]
Since callbacks are a new feature and limited to this API, we are free to add strict requirements on their signature. However, this should
1. be _very_ explicitly documented and
2. be covered by the API contract, so can't be changed after the release.

This also includes the return value (introduced in your other PR).

(I know you are fully aware of this, but I thought it'd merit noting down.)

**Comment by fsouza** [comment]
> The arguments in a callback shouldn't ever have been relied on.

I'm not sure I understand what you mean. This PR is adding an argument to a callback, shouldn't users to rely on that?

Btw, I understand that anything on master is unstable, was just calling out that this is also a breaking change in Lua, since the PR description claims it's a breaking change only in VimScript.

**Comment by gpanders** [comment]
@fsouza I will mention that in the breaking changes issue when/if this is merged. Thanks for pointing that out.

**Comment by clason** [comment]
> I'm not sure I understand what you mean. 

There's been no statement that the callback gets passed _anything_  -- in fact, currently it isn't, ever, so your example is not functional (in the sense that it only ever executes the `or` codepath).

**Comment by lewis6991** [comment]
> This PR is adding an argument to a callback, shouldn't users to rely on that?

Yes, once this PR is merged, users can then rely on the arguments. My comment referenced the past tense. Users should not have relied on arguments passed to callbacks because it was not defined what they would be, so it's not quite a breaking change in that regard. With this PR those arguments will be defined.

**Comment by fsouza** [comment]
> Yes, once this PR is merged, users can then rely on the arguments. My comment referenced the past tense. Users should not have relied on arguments passed to callbacks because it was not defined what they would be, so it's not quite a breaking change in that regard. With this PR those arguments will be defined.

So if a second parameter is introduced, would you make the same argument? I guess you're saying that you don't consider adding a parameter to a callback a breaking change? Fair enough.

**Comment by max397574** [comment]
I think it's not a breaking change because it was never intended before that you use that parameter. It also wasn't documented.
At least that's how I understood it.

**Comment by clason** [comment]
> > Yes, once this PR is merged, users can then rely on the arguments. My comment referenced the past tense. Users should not have relied on arguments passed to callbacks because it was not defined what they would be, so it's not quite a breaking change in that regard. With this PR those arguments will be defined.
> 
> So if a second parameter is introduced, would you make the same argument? I guess you're saying that you don't consider adding a parameter to a callback a breaking change? Fair enough.

Note that your example is pointless, since you cannot "wish for" arguments to be passed -- only those that are hardcoded in the API are (which before this PR were none).

**Comment by fsouza** [comment]
> Note that your example is pointless, since you cannot "wish for" arguments to be passed -- only those that are hardcoded in the API are (which before this PR were none).

The snippet I shared is adapted from [real code](https://github.com/fsouza/dotfiles/blob/93fe3cf9becc9dacbcc3e1137e3fc6eb57596712/nvim/lua/fsouza/plugin/editorconfig.fnl#L93-L98).

The fact that a function is used as a callback doesn't mean that that's the sole use of that function, at least that's how I understood it, but it seems that my understanding was incorrect, which is fine, I've adapted the code.

**Comment by gpanders** [comment]
The degree to which this is a breaking change is not really relevant. Can we please keep the discussion to the changes introduced in the PR?

**Comment by lewis6991** [comment]
Arguments to callbacks were never documented. That shouldn't be implied to mean the callback won't pass any arguments.

After this PR, it will be documented so there will be a stronger argument for saying additional parameters would be a breaking change.

I don't think anyone should rely on callbacks missing certain undefined parameters unless the API explicitly states this can be done. This also goes for named keys in option tables which are never considered a breaking change when new options are added.

File: src/nvim/api/autocmd.c [hunk 1]
```diff
@@ -315,8 +315,13 @@ Array nvim_get_autocmds(Dict(get_autocmds) *opts, Error *err)
 ///
 /// @param opts Optional Parameters:
 ///         - callback: (string|function)
-///             - (string): The name of the viml function to execute when triggering this autocmd
-///             - (function): The lua function to execute when triggering this autocmd
+///             - (string): The name of the Vimscript function to execute when this autocmd is
+///                         triggered.
+///             - (function): Lua function to call when triggering this autocmd.
+///             - The function is passed a dict or table with the keys "match", "file", and "buf"
```
Comment (lewis6991) [inline src/nvim/api/autocmd.c:321]: Say "dict" or "table", but not "dict or table" as it implies one or the other.

File: src/nvim/api/autocmd.c [hunk 2]
```diff
@@ -315,8 +315,13 @@ Array nvim_get_autocmds(Dict(get_autocmds) *opts, Error *err)
...
+///               with values corresponding to the expanded values of |<amatch>|, |<afile>|, and
+///               |<abuf>|, respectively. If the callback function returns |v:true|, then the
```
Comment (lewis6991) [inline src/nvim/api/autocmd.c:323]: I'd prefer if we minimized vimscript references in the API docs. `true` is sufficient here.

Comment (gpanders) [reply src/nvim/api/autocmd.c:323]: The use of `v:true` was intentional: Vimscript functions must return `v:true` explicitly, not any “true” value. Saying it must return `true` is ambiguous, because in Vimscript `1` is true, but does not remove the autocommand.

It is less ambiguous to infer that `v:true` == `true` in Lua than the other way around.

Comment (lewis6991) [reply src/nvim/api/autocmd.c:323]: I feel this should be more explicit then. E.g. "note: for vimscript v:true must be returned".

Comment (lewis6991) [reply src/nvim/api/autocmd.c:323]: To add, nowhere else in the API docs do we make this distinction.

Comment (clason) [reply src/nvim/api/autocmd.c:323]: Never hurts to be explicit.

Comment (lewis6991) [reply src/nvim/api/autocmd.c:323]: Do we not also need to do this for `nvim_buf_attach` and `nvim_set_decoration_provider`?

Comment (gpanders) [reply src/nvim/api/autocmd.c:323]: Both of those only take Lua callbacks, so it's not an issue there.

**Comment by gpanders** [comment]
> Since callbacks are a new feature and limited to this API, we are free to add strict requirements on their signature. However, this should
> 
> 1. be _very_ explicitly documented and
> 2. be covered by the API contract, so can't be changed after the release.
> 
> This also includes the return value (introduced in your other PR).
> 
> (I know you are fully aware of this, but I thought it'd merit noting down.)

Given this, are there any other arguments we should add? I would also propose the actual event that triggered the autocommand.

**Comment by lewis6991** [comment]
Event and string that matched the pattern.

**Comment by muniter** [comment]
Very cool feature 👍

I think we should reuse the allocation, leave the responsibility of copying the object to the callback if it wants to store it/use it later, etc. Analogous to `v:event`.

About extra keys I think event name is a good idea.

**Comment by gpanders** [comment]
Actually now that you mention `v:event`, I think that would be a better place to put the current event name.

File: src/nvim/autocmd.c [hunk 3]
```diff
@@ -2023,15 +2023,22 @@ char_u *getnextac(int c, void *cookie, int indent, bool do_concat)
   }
 
   if (ac->exec.type == CALLABLE_CB) {
+    dict_T *dict = tv_dict_alloc();
```
Comment (muniter) [inline src/nvim/autocmd.c:2026]: @gpanders What I'm trying to say is that for every autocmd that matches and is a callback we are allocating memory for the args, providing a unique dictonary to the callback. Perhaps is better to have the memory allocated previously, have only one dictionary and with every pass of the args just rewrite the values that should be on the ditcionary.

If the callback wants to store it, use it later, etc. They should call `deepcopy`.

Comment (lewis6991) [reply src/nvim/autocmd.c:2026]: Doesn't Lua create copies anyway?

**Comment by lewis6991** [comment]
> Actually now that you mention v:event, I think that would be a better place to put the current event name.

How come? We don't generally use the `v:` space in the API. I would prefer the event to be an argument.

File: src/nvim/autocmd.c (hunk 3 above)
Comment (muniter) [reply src/nvim/autocmd.c:2026]: It would require extra glue code, to have a global that we update instead of converting the `typval_T` to a new lua table. Perhaps not worth all the effort? Feel free to dismiss 😄

**Comment by clason** [comment]
> > Actually now that you mention v:event, I think that would be a better place to put the current event name.
> 
> How come? We don't generally use the `v:` space in the API. I would prefer the event to be an argument.

If we want to avoid that, we should conversely expose the contents of `v:event` to the callback as an argument -- depending on the event, those are quite important...

File: src/nvim/autocmd.c (hunk 3 above)
Comment (lewis6991) [reply src/nvim/autocmd.c:2026]: But would it? Whenever you pass anything into Lua, the C API will create copies (AFAIK), so we don't have to worry about using deepcopy or anything, users just need to keep a ref to the table and the Lua-C interface will just figure it out. This typeval stuff is just an additional layer to that.

**Comment by muniter** [comment]
Honestly `v:event` feels like a huge drawer. Many things don't apply to many events.

A better interface IMHO would be to have a "data" key in the passed dictionary in which we pass **relevant event attributes**. Example if it's a popup menu event we pass `col, row, size`, if it's a terminal event we pass `status`, and so on. All this of course in a **follow up** in which we start segmenting this things and properly documenting what data passes each event.

This design would have the extra that User events could become more rich, allow for better IPC (Inter plugin communication lol). They could pass some data that the autocommand could receive. Example:

```lua
nvim_do_autocmd("User FugitiveObject", {
  pattern = "eiojfoiauji234"
  data = {
    repo = "neovim/neovim",
    commiter = "someone",
  }
}
```

**Comment by gpanders** [comment]
> > Actually now that you mention v:event, I think that would be a better place to put the current event name.
> 
> How come? We don't generally use the `v:` space in the API. I would prefer the event to be an argument.

I mean more generally, for *all* autocommands, expose a `name` field in `v:event` for the name of the event that is running. So this can be used in API callbacks but also in all autocommand execution.

The description of `v:event` is "Dictionary of event data for the current autocommand." IMO it's surprising that the name of the event is not *already* in there.

**Comment by lewis6991** [comment]
`v:event` is an anti-pattern, and we shouldn't encourage using it. We should just be passing this info directly to the callback.

**Comment by gpanders** [comment]
How is it an anti-pattern? If we expose it only to the callback then autocommands that don't use the API don't have access to that information.

**Comment by lewis6991** [comment]
> How is it an anti-pattern? 

Because it is using a global table to access what is essentially localised information. We should minimize inheriting and relying on weird vim-isms in the API. As far as I'm concerned, that kind of stuff can be considered legacy.

Add the event name`v:event` too, if you feel the need to support and develop the vimscript domain, but I don't think that should be done here.

File: src/nvim/autocmd.c (hunk 3 above)
Comment (muniter) [reply src/nvim/autocmd.c:2026]: You are right, anything we push into the stack is going to be copied anyway in this case mainly strings if they are not there already.

But I believe in the C side we can avoid allocating a dict every time we are going to pass args and just reuse the same one.

**Comment by lewis6991** [comment]
Sorry, I've failed to recognize that these callbacks are Lua only and not available to RPC plugins. In that case, adding the event name to `v:event` seems like it is worth doing. However, for anything we do with Lua callbacks, we should aim to provide as much as possible locally via arguments like we already do for other callbacks. This will make async code easier to write since users don't have to worry about the event information changing between async calls.

File: src/nvim/autocmd.c (hunk 3 above)
Comment (gpanders) [reply src/nvim/autocmd.c]: I tried this in e9d9caada27913defa0ed167a6407406e29d7563. It took a few other code changes because static dicts aren't really something that has been used before. But I do think this might be worthwhile because autocommand callbacks can potentially fire *a lot* and very quickly, so minimizing memory allocations is a worthwhile goal (there is still required allocations for adding the items to the dict, however). It's going through CI now.

**Comment by muniter** [comment]
**TLDR:** I think that we should only pass the autocommand id and group as table keys, everything extra should go under a data table and ideally should be coherent with the event.

---
I have a few more thought on this feature. And I think we should be careful so we can keep cohesion on what we pass to autocommands.

I think we all agree in that we should pass a table, but I believe "top" keys on the table should be things that all autocommands have in common, like it's id and group. And we should put autocommand **dependant** keys in a nested table.

This is because although the vaules of buf, match and file make sense **for many** autocommands they do not for many others. Here's a quick list of commands in which they are not really cohesive with the event semantics.

`ChanInfo, ChanOpen, CmdUndefined, CmdlineChanged, CmdlineEnter, CmdlineLeave, ColorSchemePre, CompleteDonePre, ExitPre, FuncUndefined, RemoteReply, Signal, SpellFileMissing, etc.`

The ideal thing is that each autocommand defines the data they want in the table, this is of course a considerable refactor that *is out of scope from this PR*, but decisions made in this PR may hinder our ability to have a more cohesive autocommand API.

Examples:

Args for ChanInfo
```lua
{
 id = 129,  -- autocommmand id
 group = 3, -- group id
 data = {
    id = 5,  -- channel id
    argv = { "ls", "-la" },
    mode = "bytes"
    ...
 }
}
```

Args for BufEnter
```lua
{
 id = 105,  -- autocommmand id
 group = 5, -- group id
 data = {
    buffer = 7,
    match = "*.lua",
    file = "/home/user/.config/nvim/init.lua",
   ...
 }
}
```

Args for TextYankPost
```lua
{
 id = 9,  -- autocommmand id
 group = 4, -- group id
 data = {
    buffer = 7,
    file = "/home/user/.config/nvim/init.lua",
    inclusive = true,
    operator = "y",
    regcontents = "vim.wo.conceallevel\nvim.wo",
    regname = "a",
    regtype = "v",
    visual = false
    ...
 }
}
```

Something like this will deprecate v:event all together, allow for **User defined commands** to also provide [their own data](https://github.com/neovim/neovim/pull/17786#issuecomment-1073310655), etc.

---

**Comment by gpanders** [comment]
Ok I just pushed version 2.0. This uses API objects instead of typvals and only supports Lua callbacks. I don't really think we need to support Vimscript callbacks, but if others feel strongly we can make the necessary modifications. I also put in the skeleton of the structure @muniter proposed: `id` and `group` at the top level of the args table with a `data` subtable that, for now, contains `match`, `file`, and `id`. We can eventually extend this to contain event-specific data.

File: src/nvim/autocmd.c [hunk 4]
```diff
@@ -1964,6 +1964,49 @@ void auto_next_pat(AutoPatCmd *apc, int stop_at_last)
...
+    Dictionary d = ARRAY_DICT_INIT;
+    PUT(d, "id", INTEGER_OBJ(ac->id));
+    PUT(d, "data", DICTIONARY_OBJ(data));
+    switch (ap->group) {
+    case AUGROUP_ERROR:
+      abort();  // unreachable
+    case AUGROUP_DEFAULT:
+    case AUGROUP_ALL:
+    case AUGROUP_DELETED:
+      PUT(d, "group", CSTR_TO_OBJ("NONE"));
```
Comment (lewis6991) [inline src/nvim/autocmd.c:1986]: This means group will be either a string or an integer. Wouldn't it be better to just use `0` or `nil` here? `nil` would allow code like: 

```lua
if group then
  ... do something with group
end
```

File: src/nvim/lua/executor.c [hunk 5]
```diff
@@ -1268,14 +1268,14 @@ int nlua_source_using_linegetter(LineGetter fgetline, void *cookie, char *name)
...
 /// @param[in]  argcount Count of typval arguments
 /// @param[in]  argvars Typval Arguments
 /// @param[out] rettv The return value from the called function.
-int typval_exec_lua_callable(lua_State *lstate, LuaCallable lua_cb, int argcount, typval_T *argvars,
-                             typval_T *rettv)
+int typval_exec_lua_callable(LuaCallable lua_cb, int argcount, typval_T *argvars, typval_T *rettv)
 {
+  lua_State *const lstate = global_lstate;
+
   LuaRef cb = lua_cb.func_ref;
```
Comment (lewis6991) [inline src/nvim/lua/executor.c:1279]: The changes in this file don't seem necessary and don't look to improve things?

Comment (gpanders) [reply src/nvim/lua/executor.c:1279]: Oops yea these shouldn’t have been included

**Review by lewis6991 — APPROVED** [review]

**Review by muniter — COMMENTED** [review]
Nice, 👍 on just Lua support, also the event name is a piece of information that all autocommmands share. I think is a good candidate to be at the top of the table. And seems useful for autocommmands that have multiple events.

**Comment by clason** [comment]
@gpanders can you rebase on @muniter's doc changes?

**Comment by lewis6991** [comment]
Now that the conditional fields have been placed in a nested table, could we change the top level fields to be positional parameters of the callback?

**Comment by clason** [comment]
@gpanders friendly ping? This is a blocker for 0.7.

**Comment by gpanders** [comment]
What is the benefit of making the top level args positional? That locks us in to that format, and we can’t add positional arguments once it’s released. And I feel like I will never remember the order of the positional args (“Is it id or group first? Better check `:h nvim_create_autocmd` again…”). In contrast if everything is a table we can add new top-level keys if we want, and order is not a concern.

**Comment by lewis6991** [comment]
Honestly it's just preference, but we already have a well established convention of using positional arguments for callbacks. And your argument is pretty fundamental about function arguments in their entirety, not specifically how they are used in this PR. Is there a good reason why we should deviate here?

Positional arguments don't limit us from adding new ones in the future (I don't think, someone correct me if I'm wrong) as long as the number of arguments isn't strictly defined.

**Comment by clason** [comment]
I too would prefer if we followed a standard paradigm for function arguments (not only here but throughout the API); in most cases recently, this was "mandatory arguments as positional, followed by a single `opts` table for optional arguments". (I personally don't like nested tables, and the "single nested table so we can mock keyword argument style" feels a bit too Lua-magical for a Neovim API to me.)

And the API contract allows adding new options as long as it breaks no valid call with the old signature (which would be difficult if we already fix an optional table here).

File: src/nvim/autocmd.c [hunk 6]
```diff
@@ -2005,6 +2005,52 @@ void auto_next_pat(AutoPatCmd *apc, int stop_at_last)
   }
 }
 
+static bool call_autocmd_callback(const AutoCmd *ac, const AutoPatCmd *apc)
+{
+  bool ret = false;
+  Callback callback = ac->exec.callable.cb;
+  if (callback.type == kCallbackLua) {
+    Dictionary data = ARRAY_DICT_INIT;
+    PUT(data, "match", CSTR_TO_OBJ((char *)autocmd_match));
```
Comment (bfredl) [inline src/nvim/autocmd.c:2014]: I think match is a fundamental property of the fired event, and weird to have stuffed away in a nested `"data"` dict

Comment (bfredl) [reply src/nvim/autocmd.c:2014]: .. which doesn't effect what I just said in the slightest? Please comment if you have anything useful to add.

File: src/nvim/eval.c [hunk 7]
```diff
@@ -7248,15 +7247,6 @@ bool callback_call(Callback *const callback, const int argcount_in, typval_T *co
     name = partial_name(partial);
     break;
 
-  case kCallbackLua:
```
Comment (bfredl) [inline src/nvim/eval.c:7251]: this code should be kept for future use

**Review by bfredl — APPROVED** [review]
otherwise lgtm
